### PR TITLE
fix(dropdown): various improvements

### PR DESCRIPTION
**Pull request by theJohnnyMe**
**Describe pull-request**  
Various fixes and improvements for Dropdown component

**Solving issue**  
Fixes:

1. [AB#1675](https://dev.azure.com/scaniadigitaldesignsystem/3d9eb9c6-0045-473d-bc1d-e842dd779200/_workitems/edit/1675)
2. [AB#1673](https://dev.azure.com/scaniadigitaldesignsystem/3d9eb9c6-0045-473d-bc1d-e842dd779200/_workitems/edit/1673)
3. [AB#1358](https://dev.azure.com/scaniadigitaldesignsystem/3d9eb9c6-0045-473d-bc1d-e842dd779200/_workitems/edit/1358), #270 
4. [AB#1386](https://dev.azure.com/scaniadigitaldesignsystem/3d9eb9c6-0045-473d-bc1d-e842dd779200/_workitems/edit/1386), #297 
5. #304 
6. #284 
7. Propagate selected-value and selected-text on the top of the parent as attributes.
8. [AB#1669](https://dev.azure.com/scaniadigitaldesignsystem/3d9eb9c6-0045-473d-bc1d-e842dd779200/_workitems/edit/1669)

**How to test**  
1. Go to Storybook
2. Check ticket per ticket from "Fixes" if it is still valid

### CODE REVIEW

**Review by alexlioliopoulos — CHANGES_REQUESTED** [review]
First of all, looks great! Love the autopositioning too.

A couple of comments so far:
- There is no focused state while the dropdown is expanded. Meaning that you cannot make selections by only using your keyboard.
- The label-inside variant is not working as intended. The label is not remaining inside the dropdown after the selection. Have a look at the current StoryBook example, or the design files for more info on that. 
image

**Review by areblk — CHANGES_REQUESTED** [review]
## Issues:
1. Typo in `<sdds-dropdown-option value="option-1">Stockhol & Stockholm</sdds-dropdown-option>` should be `<sdds-dropdown-option value="option-1">Stockholm & Stockholm</sdds-dropdown-option>` or something more abstract. I have a suggestion for special chars regression testing under **suggestions** 
2. AB#1673 AB#1443 Something weird with error message-story. Error-story presents correctly, but changing error control true/false the icon is removed. Need to go to another story and refresh browser for icon to appear again.
3. (*Alex also mentioned this*) #297 Label inside-variant should display label in top left corner of dropdown when an option is selected, see https://dw1u92ia6bs84.cloudfront.net/Images/label_inside_5709281b63.png
4. (*Alex also mentioned this*) It's possible to focus and open dropdown with keyboard {enter} but not to toggle or select an option `<sdds-dropdown-option>`. For this reason, I'm not able to verify focused state style.
5. AB#1669 (no.3) clicking outside filter dropdown doesn't clear what is typed but not selected. This is how I understood Elias feedback, but **not sure what our requirements for this feature are.**


## Not tested (not sure how):
- #304 as it depends on autocomplete data
- Not able to verify AB#1669 (no.1) because of storybook layout

## Fixed:
- AB#1675 AB#957 options with '&' in value doesn't change on select. Also tested with other special chars without failure.
- AB#1358 sdds-dropdown doesn't overlay side menu on small screens. To test this, I used the code example from 'create table' in #270 and set viewport to xs. The table and dropdown is hidden behind side menu. Although, not sure of what component caused the core issue.
- AB#1386 AB#1394 #284 Icon flips on dropdown expansion on all stories except **Native select**
- AB#1669 (no.2) : message when query resulting in 0 results 
- AB#1669 (no.4) Divider colour aligns dropdown helper type ("validity") (**EDIT:** This was under the wrong header, it works)**


## Suggesstion:
7. Add sdds-dropdown-option values displaying special chars to keep track of regression.

```
<sdds-dropdown-option value="alpha-values-1">A B C D E F G H I J K L M N O P Q R S T U V W X Y Z Å Ä Ö</sdds-dropdown-option>
<sdds-dropdown-option value="alpha-values-2">a b c d e f g h i j k l m n o p q r s t u v w x y z å ä ö</sdds-dropdown-option>
<sdds-dropdown-option value="num-values">0 1 2 3 4 5 6 7 8 9</sdds-dropdown-option>
<sdds-dropdown-option value="special-chars-1">! " # $ % & ' ( ) * + , - . /</sdds-dropdown-option>
<sdds-dropdown-option value="special-chars-2">: ; < = > ? @</sdds-dropdown-option>
<sdds-dropdown-option value="special-chars-3">[ \ ] ^ _ ` { | } ~</sdds-dropdown-option>
```

8. To test issues like AB#1358 / #270, what about refactoring the fullpage story so that page content can be set in control? Then we can test components together with header and side menu to notice if there are issues with z-index. Also, being able to set page content in control would make it possible to compare master branch to changes in PR. The whole fullpage-story is a bit broken now, so this would be a task of its own.
9. Make Extra dropdown options available also for Filter story

## Note:
- Storybook is not great for testing individual components on mobile devices (not possible to enter full screen mode and test actual mobile device behaviour) or smaller viewports. When having a short viewport, the dropdown menu expands outside the viewport. Not sure if this is a Storybook issue or a component issue. Testing it through storybook on my mobile it looked fine.

scrollable container too high

**Comment by areblk** [comment]
@theJohnnyMe On the topic of keyboard accessible dropdown, there's a js cue for allowing options to be activated on enter/return key. See 'Building keyboard accessibility back in' on https://developer.mozilla.org/en-US/docs/Learn/Accessibility/HTML

**Comment by theJohnnyMe** [comment]
> @theJohnnyMe On the topic of keyboard accessible dropdown, there's a js cue for allowing options to be activated on enter/return key. See 'Building keyboard accessibility back in' on https://developer.mozilla.org/en-US/docs/Learn/Accessibility/HTML

Aware of it, but the issue here is that dropdown menu is a shadowDOM element so it is tricky to access it. If I find some workaround I will try to apply it.

**Comment by theJohnnyMe** [comment]
@areblk and @alexlioliopoulos  - Please check the label inside again, I pushed some fixes.

**Comment by alexlioliopoulos** [comment]
Is there a preview that we can check?

**Comment by areblk** [comment]
> Is there a preview that we can check?

same link applies, https://pr-306.d1k9aiuqofg0s0.amplifyapp.com/ @alexlioliopoulos

**Comment by alexlioliopoulos** [comment]
Ah, thanks! The page was cached and was not showing me the new changes. I had to hard refresh!

**Comment by areblk** [comment]
> Please check the label inside again, I pushed some fixes.

Label inside applies when option is selected. 

Although, when the option takes up two lines, the label is pushed upwards. 
image

Not sure what the expected result is (@alexlioliopoulos), but a suggestion would be to hide overflow and appended with '...'
image

**Comment by alexlioliopoulos** [comment]
> > Please check the label inside again, I pushed some fixes.
> 
> Label inside applies when option is selected.
> 
> Although, when the option takes up two lines, the label is pushed upwards. image
> 
> Not sure what the expected result is (@alexlioliopoulos), but a suggestion would be to hide overflow and appended with '...' image

I think that's a good idea to truncate!

**Comment by areblk** [comment]
> ## Issues:
> 1. Typo in `<sdds-dropdown-option value="option-1">Stockhol & Stockholm</sdds-dropdown-option>` should be `<sdds-dropdown-option value="option-1">Stockholm & Stockholm</sdds-dropdown-option>` or something more abstract. I have a suggestion for special chars regression testing under **suggestions**
> 2. AB#1673 AB#1443 Something weird with error message-story. Error-story presents correctly, but changing error control true/false the icon is removed. Need to go to another story and refresh browser for icon to appear again.
> 3. (_Alex also mentioned this_) #297 Label inside-variant should display label in top left corner of dropdown when an option is selected, see https://dw1u92ia6bs84.cloudfront.net/Images/label_inside_5709281b63.png
> 4. (_Alex also mentioned this_) It's possible to focus and open dropdown with keyboard {enter} but not to toggle or select an option `<sdds-dropdown-option>`. For this reason, I'm not able to verify focused state style.
> 5. AB#1669 (no.3) clicking outside filter dropdown doesn't clear what is typed but not selected. This is how I understood Elias feedback, but **not sure what our requirements for this feature are.**
> 
> ## Not tested (not sure how):
> * #304 as it depends on autocomplete data
> * Not able to verify AB#1669 (no.1) because of storybook layout
> 
> ## Fixed:
> * AB#1675 AB#957 options with '&' in value doesn't change on select. Also tested with other special chars without failure.
> * AB#1358 sdds-dropdown doesn't overlay side menu on small screens. To test this, I used the code example from 'create table' in #270 and set viewport to xs. The table and dropdown is hidden behind side menu. Although, not sure of what component caused the core issue.
> * AB#1386 AB#1394 #284 Icon flips on dropdown expansion on all stories except **Native select**
> * AB#1669 (no.2) : message when query resulting in 0 results
> * AB#1669 (no.4) Divider colour aligns dropdown helper type ("validity") (**EDIT:** This was under the wrong header, it works)
> 
> ## Suggesstion:
> 7. Add sdds-dropdown-option values displaying special chars to keep track of regression.
> 
> ```
> <sdds-dropdown-option value="alpha-values-1">A B C D E F G H I J K L M N O P Q R S T U V W X Y Z Å Ä Ö</sdds-dropdown-option>
> <sdds-dropdown-option value="alpha-values-2">a b c d e f g h i j k l m n o p q r s t u v w x y z å ä ö</sdds-dropdown-option>
> <sdds-dropdown-option value="num-values">0 1 2 3 4 5 6 7 8 9</sdds-dropdown-option>
> <sdds-dropdown-option value="special-chars-1">! " # $ % & ' ( ) * + , - . /</sdds-dropdown-option>
> <sdds-dropdown-option value="special-chars-2">: ; < = > ? @</sdds-dropdown-option>
> <sdds-dropdown-option value="special-chars-3">[ \ ] ^ _ ` { | } ~</sdds-dropdown-option>
> ```
> 
> 8. To test issues like AB#1358 / #270, what about refactoring the fullpage story so that page content can be set in control? Then we can test components together with header and side menu to notice if there are issues with z-index. Also, being able to set page content in control would make it possible to compare master branch to changes in PR. The whole fullpage-story is a bit broken now, so this would be a task of its own.
> 9. Make Extra dropdown options available also for Filter story
> 
> ## Note:
> * Storybook is not great for testing individual components on mobile devices (not possible to enter full screen mode and test actual mobile device behaviour) or smaller viewports. When having a short viewport, the dropdown menu expands outside the viewport. Not sure if this is a Storybook issue or a component issue. Testing it through storybook on my mobile it looked fine.
> 
> scrollable container too high

Status update:
## Fixed:
- 1, 2, 3, 6 (wasn't an issue), 7, 9, 10

## Issues:
- 4 (focus and select options in select)
- 5 (clear typed text when clicking outside element without selecting, **still not sure if this is something we should do** @alexlioliopoulos )
- 9 extra dropdown options aren't applying to the filter variant

## Out of scope:
- 8

**Comment by areblk** [comment]
**Still some issues with no.4:** I can tab and select dropdown and option with keyboard. When one option is selected, I have to tab through 11 (not visible) elements to focus something else. Also, when I focused the dropdown without pressing enter to open it, I had to tab through all elements to focus the next element after dropdown. Perhaps this have something to do with listeners? I think the rule should be: "Only when dropdown is open should options be focusable".

Steps to reproduce:
- Press tab, dropdown is selected.
- Press enter, dropdown opens
- Press enter, focus next option.
- Press enter, option is selected and dropdown closes. Now it's not in focus
- Press tab 11 times to focus next visible element.

Steps to reproduce:
- Press tab, dropdown is selected.
- Press tab again to focus the element after dropdown. You'll notice that there are a lot of invisible elements to be tabbed through.

**Comment by areblk** [comment]
**Still some issues with no.5:** Queries with no result clears when user focuses outside filter dropdown. It's possible to focus with keyboard but behaves a bit weird when operated through keyboard compared to using a mouse.

- To select filter dropdown with keyboard and start typing to it, I have to tab one extra time. Typing state entered with a keyboard isn't styled as a focused element (thick blue line on bottom). Available options are not shown as they do when user focuses component with mouse.
- If I focus filter dropdown with keyboard and press enter, it behaves the same way as if it had been activated with a mouse.
- Because the second behaviour is analog to user that operates with a mouse, the first behaviour is wrong, i.e.  user shouldn't be able to type by tabbing one more time.
- User shouldn't be able to select option "No results" with a keyboard (it's not able to select with a mouse)

**Newly discovered:** 
- Native select not possible to focus at all with keyboard

**Comment by theJohnnyMe** [comment]
@areblk and @alexlioliopoulos  - I am aware there are possible improvements to the accessibility via keyboard. However, I think not having any or having one where you can use a tab is a good improvement. I was trying for two hours today to figure out to disable tabIndex when the element is not opened, but I ended up in the dead-end street. 

I am sure there is a way to improve it further, but I think I would need more time to research emitters and listeners in Stencil in order to make it perfect. I say make it a ticket to work on once we fix other major issues in other components.

I have not changed anything to the native sdds dropdown so please add that as an improvement for native select, so was working only on the Stencil component.

**Review by gaonita — COMMENTED** [review]
Great work, a LOT of bug fixes and it looks good enough to me to merge with certain level of improvement. it looks like this PR is getting bigger and bigger and hard to follow or review. if the bug is not introduced because of this PR, let's fix them after this time. 

One thing I would like to try before I approve, is to test it in our demos? 
if you haven't done yet of course. and please let me know if you need assist to do it!

File: components/src/components/dropdown/dropdown.tsx [hunk 1]
```diff
@@ -187,11 +227,33 @@ export class Dropdown {
...
+              fill-rule="evenodd"
+              clip-rule="evenodd"
+              d="M16 4C9.37 4 3.996 9.374 3.996 16.004S9.371 28.007 16 28.007c6.63 0 12.004-5.374 12.004-12.003C28.004 9.374 22.629 4 16 4ZM2 16.004c0-7.732 6.268-14 14-14s14 6.268 14 14-6.268 14-14 14-14-6.268-14-14Z"
+              fill="currentColor"
+            />
+            <path
+              d="M14.803 14.47V10h2.376v4.47l-.352 4.295h-1.672l-.352-4.295Zm-.053 5.632h2.5v2.394h-2.5v-2.394Z"
+              fill="currentColor"
+            />
+          </svg>
```
[the remark this answers is not in the document]
Comment (theJohnnyMe) [reply components/src/components/dropdown/dropdown.tsx:269]: Here I preferred to inject SVG directly (downloaded from sdds-website) as I would like to avoid dependency on the icons package as we might refactor it or remove it. Also, the icon here is not subject of change compare to the one in button where the user might use a different one. Those are my 2 cents :D

[the remark this answers is not in the document]
Comment (theJohnnyMe) [reply components/src/components/dropdown/dropdown.tsx:269]: I would test it next week in demos to make sure nothing old is broken.

[the remark this answers is not in the document]
Comment (alexlioliopoulos) [reply components/src/components/dropdown/dropdown.tsx:269]: Our icon package does not include icons, so it wouldn't work here. Right?
I was thinking though, wouldn't it be better to consume the icon via our CDN?

[the remark this answers is not in the document]
Comment (theJohnnyMe) [reply components/src/components/dropdown/dropdown.tsx:269]: @alexlioliopoulos  - Yes, our icon package does not include new icons added to the sdds-website. It is more stable to rely on inline SVG in components like this as it comes built-in with component, you do not need internet connection to have it loaded. Using CDN, in an edge case, might cause text in the dropdown to appear while SVG appears a second later or so.

[the remark this answers is not in the document]
Comment (alexlioliopoulos) [reply components/src/components/dropdown/dropdown.tsx:269]: > @alexlioliopoulos - Yes, our icon package does not include new icons added to the sdds-website. It is more stable to rely on inline SVG in components like this as it comes built-in with component, you do not need internet connection to have it loaded. Using CDN, in an edge case, might cause text in the dropdown to appear while SVG appears a second later or so.

Alright! I was thinking about performance issues, but on a second thought, this does not apply in this case since the icon will be loaded no matter what. Or? I am thinking that this is actually not a problem as long as the component could be loaded on its own, and not all components need to be loaded just to use the one, we should be alright. However, in our current solution, all components need to be installed. Any thoughts on that?

**Review by alexlioliopoulos — APPROVED** [review]

File: components/src/components/dropdown/dropdown.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (theJohnnyMe) [reply components/src/components/dropdown/dropdown.tsx:269]: Yes, in our current solution when a user installs @scania/components he/she gets access to all components that we have in the system. If you take a look at the [Carbon design system]( https://www.carbondesignsystem.com/developing/react-tutorial/step-1#install-carbon), you can see they also offer to install all the components. 

At first, offering users to install only components they need might sound like a neat solution, but I am afraid it will complicate the integration of those components as well. With the installation of every component, I am afraid,  another copy of some stencil support library files will be shipped too. So it won't be a good way either. I think what we are doing now is a good way. We can see what can we do with the icons package as that one generates a lot of files when installed and maybe we can optimize it and refactor it so it serves only new files.

**Review by areblk — APPROVED** [review]

**Review by gaonita — APPROVED** [review]
